src/Control/Effect/Reader.hs

```diff
@@ -30,35 +36,35 @@ instance Effect (Reader r) where
 --
 --   prop> run (runReader a ask) == a
 ask :: (Member (Reader r) sig, Carrier sig m) => m r
-ask = send (Ask ret)
+ask = send (Ask pure)
 
 -- | Project a function out of the current environment value.
 --
 --   prop> snd (run (runReader a (asks (applyFun f)))) == applyFun f a
 asks :: (Member (Reader r) sig, Carrier sig m) => (r -> a) -> m a
-asks f = send (Ask (ret . f))
+asks f = send (Ask (pure . f))
 
 -- | Run a computation with an environment value locally modified by the passed function.
 --
 --   prop> run (runReader a (local (applyFun f) ask)) == applyFun f a
 --   prop> run (runReader a ((,,) <$> ask <*> local (applyFun f) ask <*> ask)) == (a, applyFun f a, a)
 local :: (Member (Reader r) sig, Carrier sig m) => (r -> r) -> m a -> m a
-local f m = send (Local f m ret)
+local f m = send (Local f m pure)
 
 
 -- | Run a 'Reader' effect with the passed environment value.
 --
 --   prop> run (runReader a (pure b)) == b
-runReader :: forall r sig m a . (Carrier sig m, Monad m) => r -> Eff (ReaderC r m) a -> m a
-runReader r m = runReaderC (interpret m) r
+runReader :: r -> ReaderC r m a -> m a
+runReader r = flip MT.runReaderT r . runReaderC
 
-newtype ReaderC r m a = ReaderC { runReaderC :: r -> m a }
+newtype ReaderC r m a = ReaderC { runReaderC :: MT.ReaderT r m a }
+  deriving (Alternative, Applicative, Functor, Monad, MonadFail, MonadIO, MonadPlus, MonadTrans)
 
-instance (Carrier sig m, Monad m) => Carrier (Reader r :+: sig) (ReaderC r m) where
-  ret a = ReaderC (const (ret a))
-  eff op = ReaderC (\ r -> handleSum (eff . handleReader r runReaderC) (\case
-    Ask       k -> runReaderC (k r) r
-    Local f m k -> runReaderC m (f r) >>= flip runReaderC r . k) op)
+instance Carrier sig m => Carrier (Reader r :+: sig) (ReaderC r m) where
+  eff (L (Ask       k)) = ReaderC MT.ask >>= k
+  eff (L (Local f m k)) = ReaderC (MT.local f (runReaderC m)) >>= k
+  eff (R other)         = ReaderC (MT.ReaderT (\ r -> eff (handlePure (runReader r) other)))
```

Comment:
This is the one carrier I’ve defined in terms of a monad transformer from the `transformers` package; unlike the others, it doesn’t require users to satisfy superfluous constraints (e.g. requiring `MonadPlus`, which might not even be in scope, to use `Alternative`’s interface, like some of the other transformers in that package do).